src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,663 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    See STIR/LICENSE.txt for details
+*/
+/*!
+  \file
+  \ingroup priors
+  \brief  implementation of the stir::PLSPrior class
+
+  \author Daniel Deidda
+  \author Tsai Yu Jung
+
+*/
+
+#include "stir/recon_buildblock/PLSPrior.h"
+#include "stir/Succeeded.h"
+#include "stir/DiscretisedDensityOnCartesianGrid.h"
+#include "stir/IndexRange3D.h"
+#include "stir/IO///write_to_file.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+#include "stir/info.h"
+#include <algorithm>
+using std::min;
+using std::max;
+
+/* Pretty horrible code because we don't have an iterator of neigbhourhoods yet
+ */
+
+START_NAMESPACE_STIR
+
+template <typename elemT>
+void
+PLSPrior<elemT>::initialise_keymap()
+{
+  base_type::initialise_keymap();
+  this->parser.add_start_key("PLS Prior Parameters");
+  this->parser.add_key("only 2D", &only_2D);
+  this->parser.add_key("eta", &eta);
+  this->parser.add_key("alpha", &alpha);
+  this->parser.add_key("kappa filename", &kappa_filename);
+  this->parser.add_key("anatomical_filename", &anatomical_filename);
+  this->parser.add_key("weights", &weights);
+  this->parser.add_key("gradient filename prefix", &gradient_filename_prefix);
+  this->parser.add_stop_key("END PLS Prior Parameters");
+}
+
+
+template <typename elemT>
+bool
+PLSPrior<elemT>::post_processing()
+{
+  if (base_type::post_processing()==true)
+    return true;
+  if (kappa_filename.size() != 0)
+    this->kappa_ptr = read_from_file<DiscretisedDensity<3,elemT> >(kappa_filename);
+
+  if (anatomical_filename.size() != 0){
+    this->anatomical_sptr = read_from_file<DiscretisedDensity<3,elemT> >(anatomical_filename);}
+  else{
+      this->anatomical_sptr->fill (0);
+  }
+
+
+  if (this->anatomical_filename != "0"){
+      info(boost::format("Reading MR data '%1%'")
+           % anatomical_filename  );}
+
+  bool warn_about_even_size = false;
+
+  shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_z_sptr(this->anatomical_sptr->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_y_sptr(this->anatomical_sptr->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > anatomical_im_grad_x_sptr(this->anatomical_sptr->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > norm_sptr(this->anatomical_sptr->get_empty_copy ());
+
+  compute_image_gradient_element ((*anatomical_im_grad_z_sptr),0,*this->anatomical_sptr);
+//  std::cout<<"calc gradz ="<<std::endl;
+  compute_image_gradient_element (*anatomical_im_grad_y_sptr,1,*this->anatomical_sptr);
+  compute_image_gradient_element (*anatomical_im_grad_x_sptr,2,*this->anatomical_sptr);
+
+  this->set_anatomical_grad_sptr (shared_ptr<DiscretisedDensity<3,elemT> >(anatomical_im_grad_z_sptr->clone()),0);
```

Comment:
no need for the `clone`! You're making a copy again, but the `shared_ptr` will take care of everything. (As I said, you could have just used the class members above, but fine). just say
```C++
this->set_anatomical_grad_sptr (anatomical_im_grad_z_sptr,0);
```